### ISSUE TITLE
EOFError: end of file reached

### ISSUE THREAD

**Issue by OctopusRage** [ostinelli/apnotic#26]
when i'm trying to call : response = connection.push(notification) i got EOFError: end of file reached. Is this some kind of bug or did i make a mistake in my code / connection?

NB: i use the same code like in documentation, sry for my bad english

**Comment by mat** [comment]
I am seeing the same error, too, occasionally. 

```
EOFError(end of file reached):
  /opt/ruby-2.3.1/lib/ruby/2.3.0/openssl/buffering.rb:178:in `sysread_nonblock'
  /opt/ruby-2.3.1/lib/ruby/2.3.0/openssl/buffering.rb:178:in `read_nonblock'
  net-http2 (0.13.3) lib/net-http2/client.rb:111:in `block in socket_loop'
  net-http2 (0.13.3) lib/net-http2/client.rb:108:in `loop'
  net-http2 (0.13.3) lib/net-http2/client.rb:108:in `socket_loop'
  net-http2 (0.13.3) lib/net-http2/client.rb:93:in `block (2 levels) in ensure_open'
```

So far I couldn't find a way to reproduce it. Will keep you posted once I do.

**Comment by ostinelli** [comment]
In ruby, sockets are `IO` objects. An `EOFError` (End of File error) is raised when an operation on the `IO` object references the end of the file. In this case, the gem is trying to `sysread_nonblock` on the underlying socket which has already reached the EOF. This basically means that the socket was closed due to some exception.

This is happening in NetHttp2 (the HTTP/2 gem on which Apnotic is built) in [this line](https://github.com/ostinelli/net-http2/blob/0.13.3/lib/net-http2/client.rb#L111).

I could rescue from it specifically, something along the lines of:

```ruby
@socket_thread = Thread.new do
  begin
    socket_loop
  rescue EOFError
    # socket closed, do not raise
  rescue Exception => e
    # other error
    main_thread.raise e
  ensure
    init_vars
  end
end
```

By doing this, you wouldn't see this error anymore. My only question here is that I'm not sure if this might conceal other situations (i.e. something related to a socket being closed, but important to keep track of).

Any opinions welcome, otherwise I might just as well update NetHttp2.

**Comment by ostinelli** [comment]
@OctopusRage if instead you see this _always_, then there probably is a problem with your certificate (see #25).

**Comment by mat** [comment]
I agree putting that `rescue` into NetHTTP2 has the possibility of hiding other errors and is therefore not a great option.

On the other hand, as a user of apnotic I'd not expect such a 'low level' error bubbling up, `EOFError` does not belong into the push notification domain for my taste. I'd rather see this error presented as some error code - or if possible not at all (think retries/reconnects)- to me.

As I've no idea why this is happening I am happy to monitor my setup for a while and I'd like to just leave this issue open. Maybe someone else is seeing this one sporadically and can add information.

**Comment by ostinelli** [comment]
> On the other hand, as a user of apnotic I'd not expect such a 'low level' error bubbling up, EOFError does not belong into the push notification domain for my taste. I'd rather see this error presented as some error code - or if possible not at all (think retries/reconnects)- to me.

The connection is reset and then rebuilt even when this error bubbles up, so that's not an issue.
Agreed, I'd like too to not show this error if it's due to disconnections, however - due to the way ruby handles sockets - I don't see another way of doing it, at least for now.

**Comment by ProjectDent** [comment]
I'm seeing this every time I attempt to make a push notification using the code provided in both the "Sync pushes" example, and the "Async pushes" example.

Since people seem to have only started experiencing this today, I'm guessing it could be an intermittent issue? I'm using Ruby version 2.2.0.

**Comment by OctopusRage** [comment]
Oh i see, i just figured that i'm using development certificate, so instead of using Apnotic::Connection.new(cert_path: path, url: SANDBOX_URL), now i'm Apnotic::Connection.development(cert_path: path) and it worked like a charm.
Somehow it's hard to notice Apnotic::Connection.development in docs, so maybe you should give a bold to that part :), thx for the reply

**Comment by lhoppe** [comment]
I've been experiencing this problem since for quite a while. What happens in my case is if the connection is idle, after a while something kills the connection and then the next time the Client tries to read from the socket EOF is reached causing the error.

**Comment by ostinelli** [comment]
Just to clarify: Apnotic is programmed to raise an error when the connection times out or the connection gets closed improperly. This is by choice, so that the job manager (Sidekiq, Resque,...) can retry the job. When that happens Apnotic repairs the connection, so everything just works. 

Hence, @lhoppe can I ask you why do you see this as a problem? Thanks :)

**Comment by lhoppe** [comment]
> Just to clarify: Apnotic is programmed to raise an error when the connection gets closed improperly

The EOFError comes straight from the socket, so I'm not sure you can say that in this particular case.

@ostinelli, if you're using Sidekiq (or similar) it will retry the job all right, but it's still a crash... the only difference is in this case it's not the main thread that crashes. However, and this is why in my case this is a problem, if you're not pushing notifications in delayed jobs the whole application crashes.

**Comment by ostinelli** [comment]
> The EOFError comes straight from the socket, so I'm not sure you can say that in this particular case.

Yes I can, you can see it [here](https://github.com/ostinelli/net-http2/blob/0.13.3/lib/net-http2/client.rb#L96), and it reinitializes [here](https://github.com/ostinelli/net-http2/blob/0.13.3/lib/net-http2/client.rb#L98).

> However, and this is why in my case this is a problem, if you're not pushing notifications in delayed jobs the whole application crashes.

I don't recommend blocking the main application thread to send out push notifications anyway. That being said, if you read above you see where the culprit is; if you have suggestions on how to fix it (given how ruby sockets work) then please do share :)

**Comment by lhoppe** [comment]
Sending them synchronously isn't a problem. We receive messages through a queue which are then "turned into" push notifications and the load is simply not high enough to have any delay.

Maybe Apnotic could reconnect if the socket is closed?

**Comment by ostinelli** [comment]
> Maybe Apnotic could reconnect if the socket is closed?

Reconnection logic is something that I haven't inserted anywhere yet, because I took the approach that retrying is a dedicated queue manager's job (like Sidekiq).

Hence, I see two requests coming out of this issue:

  * Avoid having a low-level error like this one bump up. However, ruby sockets return an `EOFError` for a variety of reasons and therefore this needs to be treated carefully to avoid silencing errors that actually need to be bumped up (like invalid certificates).
  * Reconnection attempts for those use cases where Sidekiq / Resque / ... are not used.

**Comment by ostinelli** [comment]
I have released Apnotic `1.0.1` and NetHttp2 `0.14.0` to address those issues.

The current solution is as follows:

  1. NetHttp2 now raises a `SocketError` instead of the low level  `EOFError`.
  2. This error can be rescued on the main thread, so @lhoppe you can just rescue from it and hence not use a separate thread (see example test [here](https://github.com/ostinelli/net-http2/blob/0.14.0/spec/api/errors_spec.rb#L75)). Please give it a go, if the proposed solution for 2 still does not work for you, please open a new issue to discuss this specific matter.

Thank you @OctopusRage, @mat, @ProjectDent and @lhoppe for your feedback.
